soh/soh/Enhancements/randomizer/randomizer.cpp

```diff
@@ -3051,9 +3052,23 @@ void DrawRandoEditor(bool& open) {
     UIWidgets::Spacer(0);
 
     ImGui::Text("Seed");
-    if (ImGui::InputText("##RandomizerSeed", seedInputBuffer, MAX_SEED_BUFFER_SIZE, ImGuiInputTextFlags_CharsDecimal | ImGuiInputTextFlags_CallbackCharFilter, UIWidgets::TextFilters::FilterNumbers)) {
+    if (ImGui::InputText("##RandomizerSeed", seedInputBuffer, MAX_SEED_BUFFER_SIZE, ImGuiInputTextFlags_AutoSelectAll)) {
+        bool passthru = true;
         uint32_t seedInput;
-        ImGui::DataTypeApplyFromText(seedInputBuffer, ImGuiDataType_U32, &seedInput, "%u");
+
+        //Check for number-only string to pass through without hashing
+        for (char a = *seedInputBuffer; seedInputBuffer; a = *++seedInputBuffer) {
+            if (!(isdigit(a))) {
+                passthru = false;
+                break;
+            }
+        }
+        
+        if (!passthru) {
+            seedInput = boost::hash_32<std::string>{}(seedInputBuffer);
+        } else {
+            ImGui::DataTypeApplyFromText(seedInputBuffer, ImGuiDataType_U32, &seedInput, "%u");
+        }
         strncpy(seedInputBuffer, std::to_string(seedInput).c_str(), MAX_SEED_BUFFER_SIZE);
     }
     UIWidgets::Tooltip("Leaving this field blank will use a random seed value automatically\nSeed range is 0 - 4,294,967,295");
```

Comment:
This should be updated to mention the availability of strings, the character limit for both strings and numbers (see my other reply on Archez's comment), what happens if you paste in a string that goes beyond those limits, and possibly that the new seed button will only create a numeric seed with the same range as the blank field (that might be better to put on the tooltip for the new seed button though).

Reply:
![Would this suffice?](https://user-images.githubusercontent.com/49068535/217950095-d4888661-8574-463b-9e59-7e38d65d3887.png)
This is what I gathered from your feedback.

Reply:
I think we can tighten it up a little more.  For this tooltip, let's say "Supports non-numeric seed values up to 1023 characters and numeric seeds in the range 0 - 4,294,967,295.  Longer values will be truncated, and non-numeric values will be converted into a numeric seed during generation.  
Leaving this field blank will result in a random numeric seed within the above range."

Note that it's 1023 because the buffer's size of 1024 needs to reserve a space for the null terminator of the string.  Then for the New Seed button, let's make its tooltip "Generates a numeric seed in the range 0 - 4,294,967,295"

Reply:
> I think we can tighten it up a little more. For this tooltip, let's say "Supports non-numeric seed values up to 1023 characters and numeric seeds in the range 0 - 4,294,967,295.

this feels like a bad idea to me, we should just treat all entered values as strings.

Reply:
to elaborate on that idea, we should never expose the numeric seed to the user. the user should just know about a seed string, and then we can use that to generate the internal numeric seed

for random seeds, we can just generate a random string and use the same logic for string -> internal numeric seed

Reply:
> to elaborate on that idea, we should never expose the numeric seed to the user.
> 
This is what I wanted originally, but I wasn't so sure how to go about it, since it seems like the ImGui functionality happens as soon as focus is lost. Maybe we could push the hashing onto the button press.

Reply:
yeah, just have it so:
* hitting "new seed" populates the seed textbox with a random string
* hitting "clear seed" clears the seed textbox
* hitting "generate randomizer"
  * with a string in the textbox
    * takes that string and uses it to generate an internal numeric seed, does all the standard rando gen stuff
  * without a string in the textbox
    * populates the seed textbox with a random string
    * takes that string and uses it to generate an internal numeric seed, does all the standard rando gen stuff

if there's something i'm missing and this idea won't work please let me know! i'm not sure how focus loss plays into it

Reply:
My question then is within the spoiler log, what should be displayed as the seed value: the entered string, or the resulting u32 value?

If its the u32 value, then we still need a way to treat u32 entered values as numbers, and not strings (as in don't hash the u32 "number"),

If we are storing the entered string into the spoiler log, then we need proper sensitization to avoid broken JSON files. This would also mean that the value in the spoiler log may be JSON-inified, making it harder to copy/paste the seed value.

Reply:
> My question then is within the spoiler log, what should be displayed as the seed value: the entered string, or the resulting u32 value?

I'd think the entered string

> If we are storing the entered string into the spoiler log, then we need proper sanitation to avoid broken JSON files.

that sounds good to me. it does leave the question of how/when to sanitize. Trying to automatically change things gets messy (blocking typing/copy paste problems etc.), it might make more sense to follow the pattern of new password prompts (display an error message when something invalid is entered)

a simple implementation could just be an :x: with a tooltip describing max length and valid characters (this could work as a way to address https://github.com/HarbourMasters/Shipwright/issues/2449)

i figure we would also want to disable the generate seed button until the string in the textbox is valid

Reply:
I would recommend that we DON'T allow all arbitrary characters.  At most, we should just allow latin alphabet characters A-Z (upper and lower case) as well as numbers like before.  If you _really_ want to, we could also allow two more characters like _ and - so that we have 64 total characters that could be part of a valid string.  

Now for the controversial part:  we just strip out any non-conforming characters automatically.  That way, we don't have to worry about JSONifying special characters.  Users shouldn't have too much trouble NOT putting those in to begin with since the tooltip will tell them which characters are valid or they'll be pasting something in from elsewhere, and we also won't be burdening them with figuring out what's wrong with the string if we just fix it ourselves.  And they'll still know what the actual string we used is since we'll be putting the sanitized version into the spoiler file.  

Plus, if we're going to add logic to disable the "generate seed" button, then we don't even need to worry about the case from your previous reply about what happens when it's hit when there's not string in the box; we'll just make it so you can ONLY hit the generate button if there's a string in there, meaning that as far as users are concerned, any string is valid (unless we end up stripping out EVERY character).  That will make sanitization easy since we can just do that as the first step of generation, and only bail out of generation to prompt if we've stripped out everything.

Reply:
> I would recommend that we DON'T allow all arbitrary characters. At most, we should just allow latin alphabet characters A-Z (upper and lower case) as well as numbers like before. If you really want to, we could also allow two more characters like _ and - so that we have 64 total characters that could be part of a valid string.

I agree

> Now for the controversial part: we just strip out any non-conforming characters automatically.

do we do this *on paste*? on keypress? if not, what happens if someone enters `$$$$$`? or `🛹🐕`? You mentioned we could "only bail out of generation to prompt if we've stripped out everything," which implies we'd do the validation when "generate seed" is pressed. I guess the idea there is to update the value in the textbox when we hit generate, and if it ends up blank just disable the button.

I'm not sure what the best UX for this would be, I could go either way

> just make it so you can ONLY hit the generate button if there's a string in there

My gut feeling to reading this was "I don't like this", but I don't think my previous proposed solution is good either. I think only enabling the "generate seed" button when there's text in the box would make sense if we have a "manual seed entry" checkbox. When unchecked, text field is hidden and we generate a random string and put it in the spoiler. Once manual entry is checked, we disable the generate seed button until a value is entered.

Reply:
I don't like that we currently do truncation when focus is lost (thus why I opened that issue for it), so yes, I'm recommending that we don't validate until the button is pressed.  I also don't care as much if we put the validated text back in the box since it will be in the spoiler file.  In fact, we could almost treat that as a feature, since that means you could theoretically format settings strings any way you wanted and not worry about them copying/pasting in extra characters, like if it were surrounded with quotes or had newlines in it.  

In general, the flow would be as follows:
* default state:  box is empty, generate button is disabled
* text is entered in the box, the generate button lights up
* generate button is pressed, text is validated
* if there were no valid characters, empty the text box, don't proceed with generation, and somehow let the user know they didn't enter a valid string
* otherwise, proceed with generation, and optionally put the validated text back in the box.  

Also, I very much like making manual seed input optional and hidden by default, since I'd guess most people will never use it.  That means we'd also have the generate button _disabled_ as soon as the manual input box is checked (and always start with a blank text box), and re-enabled as soon as its unchecked (as well as ignore any text that was pasted into the text box if the checkbox were then unchecked prior to generation) so that the generate button is only ever disabled when manual seed input is enabled.

Reply:
> we'd also have the generate button disabled as soon as the manual input box is checked (and always start with a blank text box), and re-enabled as soon as its unchecked

I'm mostly on board with this, but i'm not sure if we should always start with a blank text box. i dont like the idea of it clearing out if someone accidentally unchecks the manual entry box

Reply:
> I don't like that we currently do truncation when focus is lost (thus why I opened that issue for it), so yes, I'm recommending that we don't validate until the button is pressed. 

I agree focus loss is a bad time to do it, but hitting the button isn't the only alternative. We could also do live validation during text entry.

Reply:
I'd be happy with live validation too, I just know it's more work to hook up, although it sounds like somebody already has for other reasons that could be used for this.  I also don't feel particularly strongly about clearing out the text box when disabling manual entry.  The main reason I mentioned it is because we should ONLY use the text box's value if the manual checkbox is checked, so as long as we do that, I'm fine preserving its value once something is in it.  At the very beginning though, we shouldn't initially populate it with something random since the whole point of having it is for people to input what THEY want.

Reply:
> At the very beginning though, we shouldn't initially populate it with something random since the whole point of having it is for people to input what THEY want

absolutely, i just think that when you can't see the text box the value should not be touched at all